### ISSUE TITLE
Start training from scratch: 2d and 3d pose with coco dataset

### ISSUE THREAD

**Issue by Kader95Sh** [ultralytics/ultralytics#5169]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am attempting to start training YOLOv8 Pose from scratch using the COCO dataset. I have created the labels exactly as you describe on your website. I'm not sure where my mistake is, which is why I'm reaching out for assistance :).

I will share the files and an example of the labels. I hope this will make it easier.

2D:
[coco_2d_kp_yolo_yaml_asTxt.txt](https://github.com/ultralytics/ultralytics/files/12770049/coco_2d_kp_yolo_yaml_asTxt.txt)

[2D_training_error.txt](https://github.com/ultralytics/ultralytics/files/12769951/2D_training_error.txt)

yolo task=pose mode=train data=./coco_2d_kp_yolo.yaml model=yolov8n-pose.yaml epochs=300 batch=16 imgsz=640 workers=8 device=\'0,1,2,3\' patience=50 

3D:
[coco_3d_kp_yolo_yaml_asTxt.txt](https://github.com/ultralytics/ultralytics/files/12770059/coco_3d_kp_yolo_yaml_asTxt.txt)

[3d_training_error.txt](https://github.com/ultralytics/ultralytics/files/12769994/3d_training_error.txt)

yolo pose train data=coco_3d_kp_yolo.yaml model=yolov8n-pose.yaml epochs=300 batch=16 imgsz=640 workers=8 device=\'0,1,2,3\' patience=50

Labels:
2D:
[000000510594.txt](https://github.com/ultralytics/ultralytics/files/12770020/000000510594.txt)

3D:
[000000510594.txt](https://github.com/ultralytics/ultralytics/files/12770026/000000510594.txt)

Thank you for your help.

**Comment by glenn-jocher** [comment]
@Kader95Sh hello,

Thank you very much for reaching out and for providing such a detailed description of your situation.

From your provided information, it's clear that you followed our guide to labeling and starting training with YOLOv8 Pose. Your command lines for training look correctly formed as well.

Without specifying the exact error messages you're encountering during training, our troubleshooting tools are limited. The training error text files you attached would have been very helpful in identifying and addressing your issues, but unfortunately, due to GitHub policy, we can't access additional files linked in posts.

Notwithstanding, I can suggest a few general points to check on when encountering issues during training:
1. Confirm that your environment meets all prerequisites and dependencies.
2. Ensure the data path in your YAML file correctly points to your dataset and labels.
3. Verify that your labels are correctly formatted and correspond with your images.
4. Check if there's enough disk space and memory available.

Once you've ensured the above are in order, try to run your training command again. If specific error messages arise, please share them, as they will significantly aid in diagnosing problems. 

Thanks for being a part of the YOLOv8 community!

Best of luck with your project,
Glenn Jocher

**Comment by Kader95Sh** [comment]
Hallo @glenn-jocher ,

Thank you for your response.

I've aligned the labels in the same order as COCO. Is that correct?

When I attempted to initiate training for 2D pose estimation from scratch, I encountered various errors. The first time, I received an error indicating that there should be 56 labels (similar to 3D), even though I'm trying to train for 2D, which should have only 39 labels. On my second attempt, I encountered the following error. 
Here's just a portion of it:
  Epoch    GPU_mem   box_loss  pose_loss  kobj_loss   cls_loss   dfl_loss  Instances       Size

  0%|          | 0/12329 [00:00<?, ?it/s]../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [138,0,0], thread: [32,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
 Epoch    GPU_mem   box_loss  pose_loss  kobj_loss   cls_loss   dfl_loss  Instances       Size

  0%|          | 0/12329 [00:00<?, ?it/s]../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [141,0,0], thread: [64,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [141,0,0], thread: [65,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [141,0,0], thread: [66,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [141,0,0], thread: [67,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [141,0,0], thread: [68,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.



With 3D i am getting different type of Error:
     Epoch    GPU_mem   box_loss  pose_loss  kobj_loss   cls_loss   dfl_loss  Instances       Size

  0%|          | 0/12329 [00:00<?, ?it/s]
      1/300     0.734G      5.787          9     0.5186       9.36      4.192          9        640:   0%|          | 0/12329 [56:45<?, ?it/s]
      1/300     0.734G      5.787          9     0.5186       9.36      4.192          9        640:   0%|          | 1/12329 [56:45<11662:11:22, 3405.57s/it]Traceback (most recent call last):
  File "/home/s/.config/Ultralytics/DDP/_temp_f25s1_4_47300741145360.py", line 9, in <module>
    trainer.train()
  File "/home/s/.local/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 195, in train
    self._do_train(world_size)
  File "/home/s/.local/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 348, in _do_train
    self.loss, self.loss_items = self.model(batch)
                                 ^^^^^^^^^^^^^^^^^
  File "/home/s/.local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/s/.local/lib/python3.11/site-packages/torch/nn/parallel/distributed.py", line 1139, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 2: 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error



Two different types of errors and i can't solve it.


Best regards

**Comment by glenn-jocher** [comment]
Hello @Kader95Sh,

Thank you for getting back with more information.

For your 2D pose estimation, your labels should be in accordance with the information found on our GitHub page. Each keypoint should be a class and the number of classes or labels should match.

The 'index out of bounds' error you encountered indicates that there might be content in your dataset that goes beyond the normal or accepted range. This might be caused by some labels or bounding boxes having coordinates that fall outside the dimensions of the respective images. Please double check your annotations to ensure all points fall within the image.

For your 3D pose estimation, the traceback shows that your module has parameters that were not used in producing the loss. You might have some additional or unused layers in your defined model that are not used during forward propagation to generate the loss. I would recommend re-checking your model configuration file and ensure that all layers are used in producing the loss.

I hope this assists you further. Please report back on your progress.

Regards,
Glenn Jocher

**Comment by Kader95Sh** [comment]
Hallo @glenn-jocher,

I tried to follow your tipps, but there is somthing i can't find cause me the errors. I am not sure if the YAML file right or not, but the labels are 100% right.

Can you take a look on my YAML file and tell me where my misstake is. i copied it from your page and using COCO-Dataset. it have to be right.

2D:
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: /scratch/HSTR-DLRL/coco_kp_2d_yolov8/  # dataset root dir
train: train  # train images (relative to 'path') 4 images
val: val  # val images (relative to 'path') 4 images
test:  test # test images (optional)

# Keypoints
kpt_shape: [17, 2]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

# Classes dictionary
names:
  0: person

Example for Labels:
COCO image: 000000000036.jpg
000000000036.txt: 
0 0.6712785862785863 0.6179453125000001 0.6457588357588357 0.726859375 0.5197505197505198 0.38125 0.5509355509355509 0.3484375 0.4885654885654886 0.3671875 0.6424116424116424 0.3546875 0.4885654885654886 0.3953125 0.738045738045738 0.5265625 0.446985446985447 0.534375 0.8461538461538461 0.771875 0.44282744282744285 0.8125 0.9251559251559252 0.9640625 0.5072765072765073 0.6984375 0.7027027027027027 0.9421875 0.5550935550935551 0.95 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 


train: 
yolo task=pose mode=train data=./coco_2d_kp_yolo.yaml model=yolov8n-pose.yaml epochs=300 batch=16 imgsz=640 workers=8 device=\'0,1,2,3\' patience=50 


Error:
RuntimeError: [Rank 1] Caught collective operation timeout: WorkNCCL(SeqNum=6, OpType=ALLREDUCE, TensorShape=[1], Timeout(ms)=10800000) ran for 10800446 milliseconds before timing out.
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 270508 closing signal SIGTERM
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 1 (pid: 270509) of binary: /work/HSTR-DLRL/Github_Tools/ultralytics/venv2/bin/python3

--------------------------------------------

3D:
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: /scratch/HSTR-DLRL/coco_kp_3d_yolov8/  # dataset root dir
train: train  # train images (relative to 'path') 4 images
val: val  # val images (relative to 'path') 4 images
test:  test # test images (optional)

# Keypoints
kpt_shape: [17, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]

# Classes dictionary
names:
  0: person


Example for Labels:
COCO image: 000000000036.jpg
000000000036.txt: 
0 0.6712785862785863 0.6179453125000001 0.6457588357588357 0.726859375 0.5197505197505198 0.38125 2.0 0.5509355509355509 0.3484375 2.0 0.4885654885654886 0.3671875 2.0 0.6424116424116424 0.3546875 2.0 0.4885654885654886 0.3953125 2.0 0.738045738045738 0.5265625 2.0 0.446985446985447 0.534375 2.0 0.8461538461538461 0.771875 2.0 0.44282744282744285 0.8125 2.0 0.9251559251559252 0.9640625 2.0 0.5072765072765073 0.6984375 2.0 0.7027027027027027 0.9421875 2.0 0.5550935550935551 0.95 2.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 

yolo pose train data=coco_3d_kp_yolo.yaml model=yolov8n-pose.yaml epochs=300 batch=16 imgsz=640 workers=8 device=\'0,1,2,3\' patience=50


Error:
Epoch    GPU_mem   box_loss  pose_loss  kobj_loss   cls_loss   dfl_loss  Instances       Size

  0%|          | 0/12329 [00:00<?, ?it/s]../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [497,0,0], thread: [32,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [497,0,0], thread: [33,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [497,0,0], thread: [34,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [497,0,0], thread: [35,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
.....
.....
....



thank you and best regards

**Comment by glenn-jocher** [comment]
Hello @Kader95Sh,

I'm glad to hear that you've verified your labels are correct. This is indeed a crucial step towards solving your issue.

Now let's dissect the issues you're encountering one by one:

1. 2D Pose Estimation: The error message you're finding here suggests that there's an issue related to your operations timing out. As a general advice for the 'RuntimeError: Caught collective operation timeout', it usually pops up when your network or GPU devices aren't able to handle the scale of your workload or there might be an issue with the underlying hardware.

It would be helpful to do a quick sanity check. Try reducing the scale of your tasks (i.e., use a lower batch size or smaller images) and see if the error persists. Also, try running your code on another device or try checking your GPU health status, if possible.

If this does not help you to resolve the error, there might be deeper issues within the code execution or data processing that might be causing the allreduce operation to wait indefinitely for data. Debugging this could require a deeper understanding of your code and environment.

2. 3D Pose Estimation: The error message you're encountering here is indicating that some indices are out of bounds. This implies that at some point in the processing pipeline, there is an attempt to access array elements that do not exist.

In your case, if you've checked your labels, please make sure to check that your train, val, and test sets (or any other data you're feeding into the model) also fit within the dimensions defined in your model configuration.

Given the symptomatic trouble of out-of-bound indexing, the issues often happen in data preprocessing or augmentation before feeding data into the network, so you may want to closely check this part of your pipeline.

Also, it's worth checking that the 'kpt_shape' parameter in your YAML file matches the actual shapes of your keypoints. An inconsistency there could lead to an 'index out of bounds' error, too.

I hope this gives you some avenues to explore in resolving your issues. Best of luck, and feel free to get back with more updates or questions.

Best regards,
Glenn Jocher

**Comment by Kader95Sh** [comment]
Hello @glenn-jocher,

I have attempted to reduce the image size and batch size while also training on fewer images, but I'm still encountering the same persistent error. Although I've successfully utilized YOLOv8 for bounding boxes and segmentation, I'm struggling to find a solution for the pose-related aspect of the project.

I'm uncertain whether the issue lies in my training methodology or if it's related to a package update. I'm currently working on a DGX-2 machine and used 4-6 V100 GPUs to train the Pose, managed through SLURM. It's worth noting that I'm simultaneously running other training tasks on the same hardware, so I believe the hardware itself is not the root cause.

Best Regards,
